pkg/arvo/lib/eth-abi.hoon

```diff
@@ -0,0 +1,559 @@
+=,  ethereum-types
+
+|%
++$  etyp  etyp:abi:ethereum
+
++$  abi-events  (pair (map @ux @tas) (map @tas (pair (list etyp) type)))
+
++$  contract
+  $:  name=@tas
+      write-functions=(map @tas function)
+      read-functions=(map @tas function)
+      events=(map @ux event)
+  ==
+
++$  function
+  $:  name=@tas
+      sol-name=@tas
+      hash=@ux
+      inputs=(list func-input)
+      outputs=(list etyp)
+  ==
++$  event
+  $:  name=@tas
+      inputs=(list event-input)
+  ==
++$  event-input  [name=@tas type=etyp indexed=?]
++$  func-input  [name=@tas type=etyp]
++$  event-input-raw  [name=@t type=@t indexed=?]
++$  func-input-raw  [name=@t type=@t]
++$  contract-raw
+  $:  name=@t
+      entries=(list entry-raw)
+  ==
++$  entry-raw
+  $%  [%function p=function-raw]
+      [%event p=event-raw]
+  ==
++$  function-raw
+  $:  name=@t
+      inputs=(list func-input-raw)
+      outputs=(list @t)
+      mut=?
+      pay=?
+  ==
++$  event-raw
+  $:  name=@t
+      inputs=(list event-input-raw)
+  ==
+
+  ++  etyp-to-type
+    |*  type=etyp
+    ^-  ^type
+    ?+  type  !!
+    ::
+        %address
+      [%atom %ux ~]
+        %uint
+      [%atom %ud ~]
+        %bool
+      [%atom %f ~]    ::
+        %int :: doesn't work with decode lib yet
+      !!
+    ::
+        %string
+      [%atom %t ~]    ::
+        %bytes
+      -:!>(octs)
+    ::
+        [%bytes-n *]
+      -:!>(*octs)
+    ::
+        [%array *]
+      [%atom %ux ~]
+    ::
+        [%array-n *]
+      [%atom %ux ~]
+    ==
+::  maybe (list ?(?(@ [@ @] (list @)) (list ?(@ [@ @] (list @))))))
+:: (list ?(?(@ [@ @]) (list ?(@ [@ @] (list @)))))
+  ++  encode-topics
+    |=  [types=(list etyp) topics=(list ?(@ (list @)))]
```

Comment:
Kinda weird how the `types` is a flat list, but it's applied against a potentially two-dimensional `topics` list.

Reply:
Yeah, a bit. The 2 dimensional list is what eth_getLogs in +request-to-json and eth-watcher's config expects.